### DISCUSSION TITLE
Instant Display of Large PDF Files: Modifying pdf-js Components for Background Downloading

### DISCUSSION THREAD

**Discussion by rocheston** [DustinBrett/daedalOS#313] [Feature Requests]
I have PDF files exceeding 200MB, and pdf-js requires downloading the entire file before displaying the first page, which is time-consuming. How can I modify components/app/pdf to instantly display the page while simultaneously downloading the file in the background?

My pdf files are linearized.

**Comment by DustinBrett** [comment]
Hi @rocheston, thanks for the question. This will be tricky for a few reasons, one of them being that the PDF app is not very efficient when loading large PDF's currently, even once it has the entire file local. I had planned at some point to add lazy loading to how it loads the pages because as it is now it can get quite slow when it's many pages. As for loading in the background or via streaming, that is another complexity as I am not sure if PDF's can be loaded intermittently, and if they can I imagine it would be something pdf.js would need to support. I don't know enough about PDF's to even say what a "linearized" means in the context of being able to load it this way. I will add this as a feature request and in the mean time if you find anything in the [pdf.js docs/repo](https://github.com/mozilla/pdf.js) about how to do this please post it here and I can give further advice on how to integrate the feature.

**Comment by DustinBrett** [comment]
I see [a bit about linearized PDF's](https://github.com/search?q=repo%3Amozilla%2Fpdf.js%20Linearized%20&type=code) so maybe it's possible. Do you have an example link of a linearized PDF I could use for testing?

**Comment by DustinBrett** [comment]
I played with this a bit last night and made some progress in lazy loading, but ran into trouble with reliability of intersection observer. Do you have a linearized PDF @rocheston? Also do you have it un-linearized so I can compare?

I think the way the code is, the file system will need to read the file totally, but we can avoid loading every page in advance which is how it is now.

**Comment by rocheston** [comment]
I made an attempt, but unfortunately, it wasn't successful. The PDF is completely downloaded to determine the total number of pages. The issue seems to stem from PDF/JS.

I tried these settings. Doesn't seem to work.
        "disableAutoFetch": true,
        "disableFontFace": false,
        "disableRange": false,
        "disableStream": true

To create linearized PDF in mac, open a PDF in Preview --> Export to PDF (Select Linearized in the save option).

**Reply by DustinBrett** [reply]
I don't have Mac or a tool to export PDF's. If you want to attach examples of before/after I can look into it further. Otherwise the best I can look into is lazy loading the pages.

**Comment by rocheston** [comment]
Here is the linearized PDF. https://rcce.cc/SQCaX

**Comment by DustinBrett** [comment]
I've added some rendering async logic.

https://github.com/DustinBrett/daedalOS/commit/65a78726a049cb37503a35d6d739e1f05b3a0deb

**Comment by rocheston** [comment]
Amazing! It's functioning perfectly now. The PDF loads seamlessly. A huge thank you, Dustin, for your incredible assistance and support. 🙏
